File: 13.roman-to-integer.py
#
# @lc app=leetcode id=13 lang=python3
#
# [13] Roman to Integer
#

# @lc code=start
class Solution:
    def romanToInt(self, s: str) -> int:
        toInt = {
            'I': 1,
            'V': 5,
            'X': 10,
            'L': 50,
            'C': 100,
            'D': 500,
            'M': 1000
        }

        sLen = len(s)

        sum = 0
        for i in range(sLen-1):
            curr = toInt[ s[i] ]
            next = toInt[ s[i+1] ]

            # (if_condition_FALSE, if_condition_TRUE)[condition]
            # sum = (sum + curr, sum - curr)[curr < next]

            # if_condition_TRUE if condition else if_condition_FALSE
            sum = sum - curr if curr < next else sum + curr
        
        sum += toInt[ s[sLen-1] ]

        return sum

        
# @lc code=end

File: 13.roman-to-integer.cpp
/*
 * @lc app=leetcode id=13 lang=cpp
 *
 * [13] Roman to Integer
 */

// @lc code=start
class Solution
{
public:
    int romanToInt(string s)
    {

        int len = s.length();

        int toInt['Z'] = {};

        toInt['I'] = 1;
        toInt['V'] = 5;
        toInt['X'] = 10;
        toInt['L'] = 50;
        toInt['C'] = 100;
        toInt['D'] = 500;
        toInt['M'] = 1000;

        int sum = 0;

        for (int i = 0; i < len - 1; i++)
        {
            int curr = toInt[s[i]], next = toInt[s[i + 1]];
            sum = (curr < next) ? (sum - curr) : (sum + curr);
        }

        sum += toInt[s[len - 1]];

        return sum;
    }
};
// @lc code=end
